src/util/address.rs

```diff
@@ -531,23 +548,21 @@ impl Address {
 
     /// Get the address type of the address.
     /// None if unknown, non-standard or related to the future witness version.
-    pub fn address_type(&self) -> Option<AddressType> {
+    pub fn address_type(&self) -> Result<AddressType, Error> {
         match self.payload {
-            Payload::PubkeyHash(_) => Some(AddressType::P2pkh),
-            Payload::ScriptHash(_) => Some(AddressType::P2sh),
+            Payload::PubkeyHash(_) => Ok(AddressType::P2pkh),
+            Payload::ScriptHash(_) => Ok(AddressType::P2sh),
             Payload::WitnessProgram {
                 version,
                 program: ref prog,
             } => {
                 // BIP-141 p2wpkh or p2wsh addresses.
-                match version {
-                    WitnessVersion::V0 => match prog.len() {
-                        20 => Some(AddressType::P2wpkh),
-                        32 => Some(AddressType::P2wsh),
-                        _ => None,
-                    },
-                    WitnessVersion::V1 if prog.len() == 32 => Some(AddressType::P2tr),
-                    _ => None,
+                match (version, prog.len()) {
+                    (WitnessVersion::V0, 20) => Ok(AddressType::P2wpkh),
+                    (WitnessVersion::V0, 32) => Ok(AddressType::P2wsh),
+                    (WitnessVersion::V0, len) => Err(Error::InvalidWitnessProgramLength(len)),
+                    (WitnessVersion::V1, 32) => Ok(AddressType::P2tr),
+                    (version, length) => Err(Error::FutureWitnessVariant { version, program_len: length }),
```

Comment:
Why is this an error? I don't like this commit either.

Reply:
In the context of address *type* it does make sense (we don't know the address type) but maybe it enourages users to reject future versions, so should be a variant of `AddressType` instead?

Reply:
@apoelstra you can't make it not a error unless we go further than just enabling P2TR addresses and will modify `AddressType` to support arbitrary future softforks - which will be a case for debate, already started in #648, #642, #643, #646...

Pls also see my explanations in https://github.com/rust-bitcoin/rust-bitcoin/pull/563#issuecomment-922424462